lib-new/Transaction.coffee

```diff
@@ -0,0 +1,169 @@
+            when @_rolledback then 'rolled back'
+            when @expiresIn <= 0 then 'expired'
+            else 'open'
+
+    cypher: (opts={}, cb) ->
+        # Check predictable error cases to provide better messaging sooner.
+        # All of these are `ClientErrors` within the `Transaction` category.
+        # http://neo4j.com/docs/stable/status-codes.html#_status_codes
+        errMsg = switch @state
+            when 'pending'
+                # This would otherwise throw a `ConcurrentRequest` error.
+                'A request within this transaction is currently in progress.
+                    Concurrent requests within a transaction are not allowed.'
+            when 'expired'
+                # This would otherwise throw an `UnknownId` error.
+                'This transaction has expired.
+                    You can get the expiration time of a transaction through its
+                    `expiresAt` (Date) and `expiresIn` (ms) properties.
+                    To prevent a transaction from expiring, execute any action
+                    or call `renew` before the transaction expires.'
+            when 'committed'
+                # This would otherwise throw an `UnknownId` error.
+                'This transaction has been committed.
+                    Transactions cannot be reused; begin a new one instead.'
+            when 'rolled back'
+                # This would otherwise throw an `UnknownId` error.
+                'This transaction has been rolled back.
+                    Transactions get automatically rolled back on any
+                    DatabaseErrors, as well as any errors during a commit.
+                    That includes auto-commit queries (`{commit: true}`).
+                    Transactions cannot be reused; begin a new one instead.'
+
+        if errMsg
+            # TODO: Should we callback this error instead? (And if so, should we
+            # `process.nextTick` that call?)
+            # I *think* that these cases are more likely to be code bugs than
+            # legitimate runtime errors, so the benefit of throwing sync'ly is
+            # fail-fast behavior, and more helpful stack traces.
+            throw new errors.ClientError errMsg
+
+        # The only state we should be in at this point is 'open'.
+        @_pending = true
+        @_db.cypher opts, (err, results) =>
+            @_pending = false
+
+            # If this transaction still exists, no state changes for us:
+            if @_id
+                return cb err, results
+
+            # Otherwise, this transaction was destroyed -- either committed or
+            # rolled back -- so update our state accordingly.
+            # Much easier to derive whether committed than whether rolled back,
+            # because commits can only happen when explicitly requested.
+            if opts.commit and not err
+                @_committed = true
+            else
+                @_rolledback = true
```

Comment:
Turns out this case only happens on database errors, *not* client ones! In fact, client errors on first transaction queries actually still create and return a new transaction!

So given that database errors are explicitly *supposed* to roll back the transaction, I'm inclined to leave this as-is. Will still add some tests.